hawkmoth/__init__.py

```diff
@@ -151,12 +130,48 @@ def _get_filenames(self):
         """
         raise NotImplementedError(self.__class__.__name__ + '._get_filenames')
 
+    def __get_all_filenames(self):
+        nfiles = 0
+
+        filenames = glob.glob(os.path.join(self.env.config.cautodoc_root, '*.[ch]'))
+        filenames += glob.glob(os.path.join(self.env.config.cautodoc_root, '**/*.[ch]'))
+
+        for filename in filenames:
+            if os.path.isfile(filename):
+                nfiles += 1
+                yield os.path.abspath(filename)
+        else:
+            if nfiles == 0:
+                self.logger.warning(f'No source files in {self.env.config.cautodoc_root}.',
+                                    location=(self.env.docname, self.lineno))
+
+    def __build_cache(self, cache, clang_args, filename):
+        # The output depends on clang args
+        key = (filename, tuple(clang_args))
+
+        if key not in cache:
+            self.env.note_dependency(filename)
+
+            docstrings, errors = parse(filename, clang_args=clang_args)
+
+            self.__display_parser_diagnostics(errors)
+            cache[key] = docstrings
+
     def run(self):
+        clang_args = self.__get_clang_args()
+        cache = self.env.temp_data.setdefault('cautodoc_cache', {})
```

Comment:
So my limited understanding of Sphinx parallel build (`sphinx-build -j N` argument) is that `self.env` is not shared during document reading i.e. each document has its own copy, and the data gets merged at a later stage.

See documentation for `parallel_read_safe` under https://www.sphinx-doc.org/en/master/extdev/index.html#extension-metadata as well as `env-merge-info` event documentation at https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-env-merge-info

I can't be certain without actually digging into Sphinx source and/or testing, but I think this implementation is still a per-document cache, not a global cache. I'd be happy to be proven wrong, but this was my conclusion also when I opted for the simple cache.

Reply:
Also see the todo example extension at https://www.sphinx-doc.org/en/master/development/tutorials/todo.html
I think the only ways to benefit from global cache is to either
1) Disable parallel build in the extension setup, so we can use the cache synchronously across the documents.
**or**
2) Switch to a three phase build where the first phase simply gathers the directives and the filenames to parse, the second phase is the `env-merge-info` merges them, and the third phase is the `doctree-resolved` event where stuff gets parsed and output generated.

Reply:
Interesting. I didn't test with parallel build indeed so I failed to notice this. I also assumed that part was fine because this is how you handled the single file cache, which apparently does not cache as much as I thought it did. Of course this needs to be looked into before going any further.

Thanks for the references. I'll see what can be done about this.